python-package/securexgboost/core.py

```diff
@@ -1582,39 +1685,97 @@ def predict(self, data, output_margin=False, ntree_limit=0, pred_leaf=False,
+                                              ctypes.c_uint(ntree_limit),
+                                              ctypes.byref(length),
+                                              ctypes.byref(preds),
+                                              c_str(username)))
+
+            # TODO(rishabh): implement this in decrypt_predictions
+            #  preds = ctypes2numpy(preds, length.value, np.float32)
+            #  if pred_leaf:
+            #      preds = preds.astype(np.int32)
+            #
+            #  nrow = data.num_row()
+            #  if preds.size != nrow and preds.size % nrow == 0:
+            #      chunk_size = int(preds.size / nrow)
+            #
+            #      if pred_interactions:
+            #          ngroup = int(chunk_size / ((data.num_col() + 1) * (data.num_col() + 1)))
+            #          if ngroup == 1:
+            #              preds = preds.reshape(nrow, data.num_col() + 1, data.num_col() + 1)
+            #          else:
+            #              preds = preds.reshape(nrow, ngroup, data.num_col() + 1, data.num_col() + 1)
+            #      elif pred_contribs:
+            #          ngroup = int(chunk_size / (data.num_col() + 1))
+            #          if ngroup == 1:
+            #              preds = preds.reshape(nrow, data.num_col() + 1)
+            #          else:
+            #              preds = preds.reshape(nrow, ngroup, data.num_col() + 1)
+            #      else:
+            #          preds = preds.reshape(nrow, chunk_size)
+        if decrypt:
+            return self.decrypt_predictions(preds, length.value)
         return preds, length.value
 
-    def save_model(self, fname):
+    # TODO(rishabh): change encrypted_preds to Python type from ctype
+    def decrypt_predictions(self, encrypted_preds, num_preds):
         """
-        Save the model to a file.
+        Decrypt encrypted predictions
+
+        Parameters
+        ----------
+        key : byte array
+            key used to encrypt client files
+        encrypted_preds : c_char_p
+            encrypted predictions
+        num_preds : int
+            number of predictions
+
+        Returns
+        -------
+        preds : numpy array
+            plaintext predictions
+        """
+        try:
+            sym_key = globals()["current_user_sym_key"]
+        except:
+            raise ValueError("User not found. Please set your username, symmetric key, and public key using `init_user()`")
+
+        # Cast arguments to proper ctypes
+        c_char_p_key = ctypes.c_char_p(sym_key)
+        size_t_num_preds = ctypes.c_size_t(num_preds)
+
+        preds = ctypes.POINTER(ctypes.c_float)()
+
+        _check_call(_LIB.decrypt_predictions(c_char_p_key, encrypted_preds, size_t_num_preds, ctypes.byref(preds)))
+
+        # Convert c pointer to numpy array
+        preds = ctypes2numpy(preds, num_preds, np.float32)
+        return preds, num_preds
```

Comment:
Do we need to return num_preds as well? The user should already have this value since they passed it in.

Reply:
Good point. Fixed.